bin/propolis-server/src/lib/vm/request_queue.rs

```diff
@@ -164,248 +214,330 @@ enum RequestDisposition {
-            allowed: AllowedRequests {
-                start,
-                migrate_as_source: RequestDisposition::Deny(
-                    RequestDeniedReason::InstanceNotActive,
-                ),
-                reboot: RequestDisposition::Deny(
-                    RequestDeniedReason::InstanceNotActive,
-                ),
-                mutate: RequestDisposition::Deny(
-                    RequestDeniedReason::InstanceNotActive,
-                ),
-                stop: RequestDisposition::Enqueue,
-            },
+            state_queue: Default::default(),
+            component_queue: Default::default(),
+
+            state: instance_state,
+            awaiting_reboot: false,
+            awaiting_migration_out: false,
+            awaiting_stop: false,
             log,
         }
     }
 
-    /// Pops the request at the front of the queue.
+    /// Pops the next request off of the queue. If the queue contains both state
+    /// change and component change requests, the next state change request is
+    /// popped first (even if it arrived later in time than the next component
+    /// change request).
     pub fn pop_front(&mut self) -> Option<ExternalRequest> {
-        self.queue.pop_front()
+        if let Some(state_change) = self.state_queue.pop_front() {
+            Some(ExternalRequest::State(state_change))
+        } else {
+            self.component_queue.pop_front().map(ExternalRequest::Component)
+        }
     }
 
     /// Indicates whether the queue is empty.
     #[cfg(test)]
     pub fn is_empty(&self) -> bool {
-        self.queue.is_empty()
+        self.state_queue.is_empty() && self.component_queue.is_empty()
     }
 
-    /// Asks to place the supplied request on the queue. If the request is
-    /// enqueued, updates the dispositions to use for future requests.
+    /// Attempts to replace the supplied `request` on the queue, returning `Ok`
+    /// if the request was accepted and an `Err` otherwise. In the latter case,
+    /// the error contains a [`RequestDeniedReason`] that describes why the
+    /// request was rejected.
     pub fn try_queue(
         &mut self,
         request: ExternalRequest,
     ) -> Result<(), RequestDeniedReason> {
-        let disposition = match request {
-            ExternalRequest::Start => self.allowed.start,
-            ExternalRequest::MigrateAsSource { .. } => {
-                self.allowed.migrate_as_source
-            }
-            ExternalRequest::Reboot => self.allowed.reboot,
-            ExternalRequest::ReconfigureCrucibleVolume { .. } => {
-                self.allowed.mutate
+        // If the queue is in a terminal state, deny the request straightaway
+        // (unless it's a stop request, which can be ignored for idempotency).
+        let disposition = if let Some(reason) = self.state.deny_reason() {
+            if matches!(
+                request,
+                ExternalRequest::State(StateChangeRequest::Stop)
+            ) {
```

Comment:
nit, feel free to ignore: seems like `ExternalRequest` could just derive `PartialEq`?

Reply:
Unfortunately the websocket connection in the `MigrateAsSource` variant gets in the way :( This match is wordy enough, though, that I think it makes sense to have a (private) `is_stop` method on `ExternalRequest`. I'll add that.